module-3/lab-supervised-learning-feature-extraction/your-code/main.ipynb

```diff
@@ -1870,11 +1870,11 @@
   },
```

Comment:
- In this one "We see that Google Play is using a shorthand for millions.", your code doesn't check if the M is in the last character just if M is in the string . This time it works because you only have this character otherwise it wouldn't work. You can do something like the following:
```Python
#Your code 

def convert_string_to_numeric(s):    
    """
    Convert a string value to numeric. If the last character of the string is `M`, obtain the 
    numeric part of the string, multiply it with 1,000,000, then return the result. Otherwise, 
    convert the string to numeric value and return the result.
  
    Args:
        s: The Reviews score in string format.
    Returns:
        The correct numeric value of the Reviews score.
    """
    if "M" in s:
        s = s.replace('M','')
        s = float(s)*1000000
        return s
    else:
        return float(s)
test_string = '4.0M'
convert_string_to_numeric(test_string) == 4000000
#A possible way of doing it 
def convert_string_to_numeric(s):
    if s[-1] == 'M':
        return float(s[:-1]) * 1000000
    else:
        return float(s)
    return np.NaN
test_string = '4.0M'
convert_string_to_numeric(test_string) == 4000000
```
- In this one "While we can convert most of the Size values to numeric in the same way we converted the Reviews values, there is one value that is impossible to convert." , other way using shape() method, https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.shape.html:
```Python
odd_value = "Varies with device"
google_play[google_play.Size == odd_value].shape[0] / google_play.shape[0]
```
- In this case "We opt to preserve these columns and remove the rows containing missing data" you don't need to specify neither axis nor how params. Just doing the following it works: 
```Python
google_missing_removed = google_play.dropna()
```
- Good one 👍 First we remove the dollar sign. Do this in the next cell by applying the str.replace function to the column to replace $ with an empty string ('').
```Python
google_missing_removed['Price Numerical'] = google_missing_removed['Price'].str.replace('$','')
```
- In this one "Our first step will be to remove all rows with missing sentiment." the same thing as above, you don't need to specify neither axis nor how params in dropna method. 
- In "Bonus", a shorter way using rename method:
```Python
google_agg = review_missing_removed.groupby('App').Sentiment.agg([positive_function, 'count']).rename({'positive_function': 'Positive', 'count': 'Total'}, axis='columns')
```
- Good one 👍 "Add a derived column to google_agg that is the ratio of the Positive and the Total columns. Call this column Positive Ratio."  I include below another way of doing the stuff:
```Python
#Your code 
google_agg['Positive Ratio'] = np.where(google_agg['Total']==0,0,google_agg['Positive']/google_agg['Total'])
#A variant of your code 
google_agg['Positive Ratio'] = np.where(google_agg['Total'] !=0, google_agg['Positive'] / google_agg['Total'], 0)
```